include/nektar_interface/projection/algorithm_types.hpp

```diff
@@ -0,0 +1,269 @@
+            Shape::template project_one_particle<nmode, T, alpha, beta>(
+                eta0, eta1, qoi, cell_dof);
+          }
+        });
+      });
+    } else {
+      return queue.submit([&](sycl::handler &cgh) {
+        cgh.parallel_for<>(range, [=](sycl::id<1> idx) {
+          auto cellx = data.cell_ids[idx];
+          auto npart = data.par_per_cell[cellx];
+
+          if (npart == 0)
+            return;
+
+          auto cell_dof = &data.dofs[data.dof_offsets[cellx]];
+          for (int part = 0; part < npart; ++part) {
+            auto xi0 = data.positions[cellx][0][part];
+            auto xi1 = data.positions[cellx][1][part];
+            auto xi2 = data.positions[cellx][2][part];
+            T eta0, eta1, eta2;
+            Shape::template loc_coord_to_loc_collapsed<T>(xi0, xi1, xi2, eta0,
+                                                          eta1, eta2);
+            auto qoi = data.input[cellx][componant][part];
+            Shape::template project_one_particle<nmode, T, alpha, beta>(
+                eta0, eta1, eta2, qoi, cell_dof);
+          }
+        });
+      });
+    }
+  }
+};
+
+// Untested on Microsoft
+// c++20 has a solution for this std::source_location::function_name
+#ifdef _MSC_VER
+#define PRETTY_FUNCTION __FUNCSIG__
+#else
+#define PRETTY_FUNCTION __PRETTY_FUNCTION__
+#endif
+
+struct ThreadPerDof {
+
+  template <int nmode, typename T, int alpha, int beta, typename Shape>
+  std::optional<sycl::event> static inline project(DeviceData<T> &data,
+                                                       int componant,
+                                                       sycl::queue &queue) {
+
+    // TODO: What to do when local_size comes back as zero
+    // 1. Theoretically can use fewer than the warp size for the block size
+    //    just be slow
+    // 2. Crash always
+    // 3. Crash in debug/warn in release <--- doing this for now, crashing will
+    //    compilcate the benchmarking in release
+    auto local_size = Private::calc_max_local_size<nmode, T, Shape>(
+        queue, Constants::preferred_block_size);
+    if (local_size == 0) {
+      fprintf(stderr,
+              "%s: This kernel uses too much local(shared) memory for your "
+              "device\n",
+              PRETTY_FUNCTION);
+      // Crash in debug version
+      assert(false);
+      // Just return empty optional in release mode
+      // Doing this so the benchmarks can run without crashing
+      // TODO: This needs a better solution I could raise an exception (blurg)
+      // or make it return std::option (more rusty)
+      return {};
+    }
+
+    std::size_t outer_size = ROUND_UP_TO_MULTIPLE(local_size, data.nrow_max);
+
+    sycl::nd_range<2> range(sycl::range<2>(data.ncells, outer_size),
+                                sycl::range<2>(1, local_size));
+    // in practice won't overflow a int
+    assert((local_size + 1) < std::numeric_limits<int>::max());
```

Comment:
I think outer_size is the one to check? CUDA backends have a habit of invisibly not running if the iteration set is too large.